### PR TITLE
Developer guide overhaul

### PR DESCRIPTION
**Pull request by tkknight**
### Description

Overhaul of the **Developers Guide** section in the documentation.  Notable changes include:

- New sections:
  - Continuous Integration (CI) Testing
  - Getting Involved
  - Running the tests
- Updated sections:
  - Contributing (github convention) now points to the docs
  - Code Formatting
  - Graphics tests
  - Pull request check list
  - Reworking of the existing git guide  (gitwash)
  - Excluded some git urls from the linkcheck to avoid hitting the domain too many times and being blocked
  - PR template (points to new PR checklist)
- Various spelling and layout updates

For the lifetime of this PR you can view the rendered html here: 
- https://iris-test-doc.readthedocs.io/en/latest/

If you wish to read the changes as rendered, I suggest starting from the **Getting Involved** section and hitting **Next** to review the entire **Developers Guide** section: 
- https://iris-test-doc.readthedocs.io/en/latest/developers_guide/contributing_getting_involved.html

---

  1. [ ] **What's new** entry primed in a separate PR: https://github.com/SciTools/iris/pull/3854

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

### CODE REVIEW

File: docs/iris/src/developers_guide/contributing_documentation.rst
Comment (rcomer) [inline docs/iris/src/developers_guide/contributing_documentation.rst:33]: I think it would be good to reinstate the info for what directory you need to be in.

**Review by rcomer — COMMENTED** [review]
Hi @tkknight, thanks for doing this.  It's a massive improvement!  I don't think I've looked at everything, but just have some comments aimed at making it a bit more idiot-proof.

File: docs/iris/src/developers_guide/contributing_pull_request_checklist.rst
Comment (tkknight) [reply docs/iris/src/developers_guide/contributing_pull_request_checklist.rst:9]: ok - I'm happy to remove the paragraph completely?

Comment (rcomer) [reply docs/iris/src/developers_guide/contributing_pull_request_checklist.rst:9]: Sounds good to me 👍

File: docs/iris/src/developers_guide/contributing_code_formatting.rst
Comment (tkknight) [reply docs/iris/src/developers_guide/contributing_code_formatting.rst:23]: Indeed.  Will add a note.

**Comment by tkknight** [comment]
Hi @rcomer.  I have addressed all your suggestions, the updated docs should be present in a few minutes (see it building: https://readthedocs.org/projects/iris-test-doc/builds/11893698/)

**Comment by rcomer** [comment]
Thanks @tkknight, all the changes look good to me.  Think this still needs a review from somebody who understands all the technical ins and outs.  In that context, I'm a nobody!

File: docs/iris/src/common_links.inc
Comment (trexfeathers) [inline docs/iris/src/common_links.inc:4]: Not something for this PR, but given its continued inclusion here, I'd just like to raise that checking the Google Group needs to be more enshrined as part of the role of a core dev. It's not something I check nearly enough.

Comment (tkknight) [reply docs/iris/src/common_links.inc:4]: I think the Google Groups usage is something we need to review - either start using it or make a statement of intent to stop using it.  I will capture this as an issue (discussion piece)

Comment (tkknight) [reply docs/iris/src/common_links.inc:4]: Discussion issue raised about google groups usage, see https://github.com/SciTools/iris/issues/3860

**Review by trexfeathers — CHANGES_REQUESTED** [review]
First of all, @tkknight and @bjlittle I just want to say what a excellent job you've done here 😍

I'm sure I'm not the only one who has in the past been cautioned against following the existing guide, for fear of it being outdated. But now I can see all the necessary knowledge - which many of us have learned in patches, with some difficulty - all combined into a single easy to follow guide - brilliant!

I've found a few minor points of grammar and readability that could be improved. Since @tkknight is enjoying some **well earned** time off I'm sure I can find someone else on the release team to push changes to this branch.

File: docs/iris/src/developers_guide/contributing_documentation.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_documentation.rst]: 07cc7eda980ce2ac1e37007c267c35e0e565c9e1

File: docs/iris/src/developers_guide/contributing_code_formatting.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_code_formatting.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/documenting/docstrings.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/docstrings.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/documenting/rest_guide.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/documenting/rest_guide.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_graphics_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_graphics_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_graphics_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_graphics_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_graphics_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_graphics_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_ci_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_ci_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_ci_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_ci_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_ci_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_ci_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

File: docs/iris/src/developers_guide/contributing_ci_tests.rst
Comment (trexfeathers) [reply docs/iris/src/developers_guide/contributing_ci_tests.rst]: 232f384223ab69017e61454bdb72d0f6aa41fa63

**Comment by bjlittle** [comment]
@trexfeathers Thanks for the pair review to see this over the line... over to you now.

**Review by trexfeathers — APPROVED** [review]

**Comment by trexfeathers** [comment]
🎉🎉

**Comment by rcomer** [comment]
Awesome work y'all!

**Pull request merged**
